graylog2-server/src/main/java/org/graylog/plugins/views/search/engine/normalization/PluggableSearchNormalization.java

```diff
@@ -0,0 +1,64 @@
+/*
+ * Copyright (C) 2020 Graylog, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the Server Side Public License, version 1,
+ * as published by MongoDB, Inc.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * Server Side Public License for more details.
+ *
+ * You should have received a copy of the Server Side Public License
+ * along with this program. If not, see
+ * <http://www.mongodb.com/licensing/server-side-public-license>.
+ */
+package org.graylog.plugins.views.search.engine.normalization;
+
+import org.graylog.plugins.views.search.Search;
+import org.graylog.plugins.views.search.permissions.SearchUser;
+import org.graylog.plugins.views.search.rest.ExecutionState;
+
+import javax.inject.Inject;
+import java.util.Collections;
+import java.util.Set;
+
+import static com.google.common.base.MoreObjects.firstNonNull;
+
+public class PluggableSearchNormalization implements SearchNormalization {
+    private final Set<SearchNormalizer> pluggableNormalizers;
+    private final Set<SearchNormalizer> postValidationNormalizers;
+
+    @Inject
+    public PluggableSearchNormalization(Set<SearchNormalizer> pluggableNormalizers,
+                                        @PostValidation Set<SearchNormalizer> postValidationNormalizers) {
+        this.pluggableNormalizers = pluggableNormalizers;
+        this.postValidationNormalizers = postValidationNormalizers;
+    }
+
+    public PluggableSearchNormalization(Set<SearchNormalizer> pluggableNormalizers) {
+        this(pluggableNormalizers, Collections.emptySet());
+    }
+
+    private Search normalize(Search search, Set<SearchNormalizer> normalizers, SearchUser searchUser, ExecutionState executionState) {
+        Search normalizedSearch = search;
+        for (SearchNormalizer searchNormalizer : normalizers) {
```

Comment:
Normalizers are a set, but each following gets a pre-processed result of normalizing, computed by the previous normalizer. Should we introduce some kind of order to prevent unpredictable bahaviour?

Reply:
I intentionally skipped that for now. I would assume that normalizers must not depend on each other. If we find out though that this is not necessarily the case, we can introduce ordering.